File: nautilus_trader/trading/controller.py
# -------------------------------------------------------------------------------------------------
#  Copyright (C) 2015-2025 Nautech Systems Pty Ltd. All rights reserved.
#  https://nautechsystems.io
#
#  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
#  You may not use this file except in compliance with the License.
#  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# -------------------------------------------------------------------------------------------------

from nautilus_trader.cache.base import CacheFacade
from nautilus_trader.common.actor import Actor
from nautilus_trader.common.component import Clock
from nautilus_trader.common.component import MessageBus
from nautilus_trader.common.config import ActorConfig
from nautilus_trader.common.config import ActorFactory
from nautilus_trader.common.config import ImportableActorConfig
from nautilus_trader.core.correctness import PyCondition
from nautilus_trader.core.message import Command
from nautilus_trader.model.identifiers import ComponentId
from nautilus_trader.model.identifiers import StrategyId
from nautilus_trader.portfolio.base import PortfolioFacade
from nautilus_trader.trading.config import ImportableStrategyConfig
from nautilus_trader.trading.config import StrategyFactory
from nautilus_trader.trading.messages import CreateActor
from nautilus_trader.trading.messages import CreateStrategy
from nautilus_trader.trading.messages import RemoveActor
from nautilus_trader.trading.messages import RemoveStrategy
from nautilus_trader.trading.messages import StartActor
from nautilus_trader.trading.messages import StartStrategy
from nautilus_trader.trading.messages import StopActor
from nautilus_trader.trading.messages import StopStrategy
from nautilus_trader.trading.strategy import Strategy
from nautilus_trader.trading.trader import Trader


class Controller(Actor):
    """
    The base class for all trader controllers.

    Parameters
    ----------
    trader : Trader
        The reference to the trader instance to control.
    config : ActorConfig, optional
        The configuration for the controller

    Raises
    ------
    TypeError
        If `config` is not of type `ActorConfig`.

    """

    def __init__(
        self,
        trader: Trader,
        config: ActorConfig | None = None,
    ) -> None:
        if config is None:
            config = ActorConfig()
        PyCondition.type(config, ActorConfig, "config")
        super().__init__(config=config)
        self._trader = trader

    def register_base(
        self,
        portfolio: PortfolioFacade,
        msgbus: MessageBus,
        cache: CacheFacade,
        clock: Clock,
    ) -> None:
        super().register_base(portfolio, msgbus, cache, clock)
        self.msgbus.register(endpoint="Controller.execute", handler=self.execute)

    def execute(self, command: Command) -> None:
        if isinstance(command, CreateActor):
            self.create_actor_from_config(command.actor_config, command.start)
        elif isinstance(command, StartActor):
            self.start_actor_from_id(command.actor_id)
        elif isinstance(command, StopActor):
            self.stop_actor_from_id(command.actor_id)
        elif isinstance(command, RemoveActor):
            self.remove_actor_from_id(command.actor_id)
        elif isinstance(command, CreateStrategy):
            self.create_strategy_from_config(command.strategy_config, command.start)
        elif isinstance(command, StartStrategy):
            self.start_strategy_from_id(command.strategy_id)
        elif isinstance(command, StopStrategy):
            self.stop_strategy_from_id(command.strategy_id)
        elif isinstance(command, RemoveStrategy):
            self.remove_strategy_from_id(command.strategy_id)

    def create_actor(self, actor: Actor, start: bool = True) -> None:
        """
        Add the given actor to the controlled trader.

        Parameters
        ----------
        actor : Actor
            The actor to add.
        start : bool, default True
            If the actor should be started immediately.

        Raises
        ------
        ValueError
            If `actor.state` is ``RUNNING`` or ``DISPOSED``.
        RuntimeError
            If `actor` is already registered with the trader.

        """
        self._trader.add_actor(actor)
        if start:
            actor.start()

    def create_strategy(self, strategy: Strategy, start: bool = True) -> None:
        """
        Add the given strategy to the controlled trader.

        Parameters
        ----------
        strategy : Strategy
            The strategy to add.
        start : bool, default True
            If the strategy should be started immediately.

        Raises
        ------
        ValueError
            If `strategy.state` is ``RUNNING`` or ``DISPOSED``.
        RuntimeError
            If `strategy` is already registered with the trader.

        """
        self._trader.add_strategy(strategy)
        if start:
            strategy.start()

    def start_actor(self, actor: Actor) -> None:
        """
        Start the given `actor`.

        Will log a warning if the actor is already ``RUNNING``.

        Raises
        ------
        ValueError
            If `actor` is not already registered with the trader.

        """
        self._trader.start_actor(actor.id)

    def start_strategy(self, strategy: Strategy) -> None:
        """
        Start the given `strategy`.

        Will log a warning if the strategy is already ``RUNNING``.

        Raises
        ------
        ValueError
            If `strategy` is not already registered with the trader.

        """
        self._trader.start_strategy(strategy.id)

    def stop_actor(self, actor: Actor) -> None:
        """
        Stop the given `actor`.

        Will log a warning if the actor is not ``RUNNING``.

        Parameters
        ----------
        actor : Actor
            The actor to stop.

        Raises
        ------
        ValueError
            If `actor` is not already registered with the trader.

        """
        self._trader.stop_actor(actor.id)

    def stop_strategy(self, strategy: Strategy) -> None:
        """
        Stop the given `strategy`.

        Will log a warning if the strategy is not ``RUNNING``.

        Parameters
        ----------
        strategy : Strategy
            The strategy to stop.

        Raises
        ------
        ValueError
            If `strategy` is not already registered with the trader.

        """
        self._trader.stop_strategy(strategy.id)

    def remove_actor(self, actor: Actor) -> None:
        """
        Remove the given `actor`.

        Will stop the actor first if state is ``RUNNING``.

        Parameters
        ----------
        actor : Actor
            The actor to remove.

        Raises
        ------
        ValueError
            If `actor` is not already registered with the trader.

        """
        if actor.id != self.id:
            self._trader.remove_actor(actor.id)

    def remove_strategy(self, strategy: Strategy) -> None:
        """
        Remove the given `strategy`.

        Will stop the strategy first if state is ``RUNNING``.

        Parameters
        ----------
        strategy : Strategy
            The strategy to remove.

        Raises
        ------
        ValueError
            If `strategy` is not already registered with the trader.

        """
        self._trader.remove_strategy(strategy.id)

    def create_actor_from_config(
        self,
        actor_config: ImportableActorConfig,
        start: bool = True,
    ) -> None:
        """
        Create the actor corresponding to `actor_config`.

        Parameters
        ----------
        actor_config : ImportableActorConfig
            The actor config of the actor to add.
        start : bool, default True
            If the actor should be started immediately.

        Raises
        ------
        ValueError
            If `actor.state` is ``RUNNING`` or ``DISPOSED``.
        RuntimeError
            If `actor` is already registered with the trader.

        """
        actor: Actor = ActorFactory.create(actor_config)
        self._trader.add_actor(actor)

        if start:
            actor.start()

    def create_strategy_from_config(
        self,
        strategy_config: ImportableStrategyConfig,
        start: bool = True,
    ) -> None:
        """
        Create the strategy corresponding to `strategy_config`.

        Parameters
        ----------
        strategy_config : ImportableStrategyConfig
            The strategy config of the strategy to add.
        start : bool, default True
            If the strategy should be started immediately.

        Raises
        ------
        ValueError
            If `strategy.state` is ``RUNNING`` or ``DISPOSED``.
        RuntimeError
            If `strategy` is already registered with the trader.

        """
        strategy: Strategy = StrategyFactory.create(strategy_config)
        self._trader.add_strategy(strategy)

        if start:
            strategy.start()

    def start_actor_from_id(self, actor_id: ComponentId) -> None:
        """
        Start the actor corresponding to `actor_id`.

        Will log a warning if the actor is already ``RUNNING``.

        Parameters
        ----------
        actor_id : ComponentId
            The ID of the actor to start.

        Raises
        ------
        ValueError
            If `actor` is not already registered with the trader.

        """
        self._trader.start_actor(actor_id)

    def start_strategy_from_id(self, strategy_id: StrategyId) -> None:
        """
        Start the strategy corresponding to `strategy_id`.

        Will log a warning if the strategy is already ``RUNNING``.

        Parameters
        ----------
        strategy_id : StrategyId
            The ID of the strategy to start.

        Raises
        ------
        ValueError
            If `strategy` is not already registered with the trader.

        """
        self._trader.start_strategy(strategy_id)

    def stop_actor_from_id(self, actor_id: ComponentId) -> None:
        """
        Stop the actor corresponding to `actor_id`.

        Will log a warning if the actor is not ``RUNNING``.

        Parameters
        ----------
        actor_id : ComponentId
            The ID of the actor to stop.

        Raises
        ------
        ValueError
            If `actor` is not already registered with the trader.

        """
        self._trader.stop_actor(actor_id)

    def stop_strategy_from_id(self, strategy_id: StrategyId) -> None:
        """
        Stop the strategy corresponding to `strategy_id`.

        Will log a warning if the strategy is not ``RUNNING``.

        Parameters
        ----------
        strategy_id : StrategyId
            The ID of the strategy to stop.

        Raises
        ------
        ValueError
            If `strategy` is not already registered with the trader.

        """
        self._trader.stop_strategy(strategy_id)

    def remove_actor_from_id(self, actor_id: ComponentId) -> None:
        """
        Remove the actor corresponding to `actor_id`.

        Will stop the actor first if state is ``RUNNING``.

        Parameters
        ----------
        actor_id : ComponentId
            The ID of the actor to remove.

        Raises
        ------
        ValueError
            If `actor` is not already registered with the trader.

        """
        if actor_id != self.id:
            self._trader.remove_actor(actor_id)

    def remove_strategy_from_id(self, strategy_id: StrategyId) -> None:
        """
        Remove the strategy corresponding to `strategy_id`.

        Will stop the strategy first if state is ``RUNNING``.

        Parameters
        ----------
        strategy_id : StrategyId
            The ID of the strategy to remove.

        Raises
        ------
        ValueError
            If `strategy` is not already registered with the trader.

        """
        self._trader.remove_strategy(strategy_id)

File: crates/system/src/messages/controller.rs
// -------------------------------------------------------------------------------------------------
//  Copyright (C) 2015-2026 Nautech Systems Pty Ltd. All rights reserved.
//  https://nautechsystems.io
//
//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
//  You may not use this file except in compliance with the License.
//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// -------------------------------------------------------------------------------------------------

use std::any::Any;

use nautilus_common::actor::data_actor::ImportableActorConfig;
use nautilus_core::{UUID4, UnixNanos};
use nautilus_model::identifiers::{ActorId, StrategyId};
use nautilus_trading::ImportableStrategyConfig;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "type")]
pub struct CreateActor {
    pub actor_config: ImportableActorConfig,
    pub start: bool,
    pub command_id: UUID4,
    pub ts_init: UnixNanos,
}

impl CreateActor {
    /// Creates a new [`CreateActor`] instance.
    #[must_use]
    pub const fn new(
        actor_config: ImportableActorConfig,
        start: bool,
        command_id: UUID4,
        ts_init: UnixNanos,
    ) -> Self {
        Self {
            actor_config,
            start,
            command_id,
            ts_init,
        }
    }

    #[must_use]
    pub fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "type")]
pub struct CreateStrategy {
    pub strategy_config: ImportableStrategyConfig,
    pub start: bool,
    pub command_id: UUID4,
    pub ts_init: UnixNanos,
}

impl CreateStrategy {
    /// Creates a new [`CreateStrategy`] instance.
    #[must_use]
    pub const fn new(
        strategy_config: ImportableStrategyConfig,
        start: bool,
        command_id: UUID4,
        ts_init: UnixNanos,
    ) -> Self {
        Self {
            strategy_config,
            start,
            command_id,
            ts_init,
        }
    }

    #[must_use]
    pub fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(tag = "type")]
pub struct StartActor {
    pub actor_id: ActorId,
    pub command_id: UUID4,
    pub ts_init: UnixNanos,
}

impl StartActor {
    /// Creates a new [`StartActor`] instance.
    #[must_use]
    pub const fn new(actor_id: ActorId, command_id: UUID4, ts_init: UnixNanos) -> Self {
        Self {
            actor_id,
            command_id,
            ts_init,
        }
    }

    #[must_use]
    pub fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(tag = "type")]
pub struct StartStrategy {
    pub strategy_id: StrategyId,
    pub command_id: UUID4,
    pub ts_init: UnixNanos,
}

impl StartStrategy {
    /// Creates a new [`StartStrategy`] instance.
    #[must_use]
    pub const fn new(strategy_id: StrategyId, command_id: UUID4, ts_init: UnixNanos) -> Self {
        Self {
            strategy_id,
            command_id,
            ts_init,
        }
    }

    #[must_use]
    pub fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(tag = "type")]
pub struct StopActor {
    pub actor_id: ActorId,
    pub command_id: UUID4,
    pub ts_init: UnixNanos,
}

impl StopActor {
    /// Creates a new [`StopActor`] instance.
    #[must_use]
    pub const fn new(actor_id: ActorId, command_id: UUID4, ts_init: UnixNanos) -> Self {
        Self {
            actor_id,
            command_id,
            ts_init,
        }
    }

    #[must_use]
    pub fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(tag = "type")]
pub struct StopStrategy {
    pub strategy_id: StrategyId,
    pub command_id: UUID4,
    pub ts_init: UnixNanos,
}

impl StopStrategy {
    /// Creates a new [`StopStrategy`] instance.
    #[must_use]
    pub const fn new(strategy_id: StrategyId, command_id: UUID4, ts_init: UnixNanos) -> Self {
        Self {
            strategy_id,
            command_id,
            ts_init,
        }
    }

    #[must_use]
    pub fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(tag = "type")]
pub struct RemoveActor {
    pub actor_id: ActorId,
    pub command_id: UUID4,
    pub ts_init: UnixNanos,
}

impl RemoveActor {
    /// Creates a new [`RemoveActor`] instance.
    #[must_use]
    pub const fn new(actor_id: ActorId, command_id: UUID4, ts_init: UnixNanos) -> Self {
        Self {
            actor_id,
            command_id,
            ts_init,
        }
    }

    #[must_use]
    pub fn as_any(&self) -> &dyn Any {
        self
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(tag = "type")]
pub struct RemoveStrategy {
    pub strategy_id: StrategyId,
    pub command_id: UUID4,
    pub ts_init: UnixNanos,
}

impl RemoveStrategy {
    /// Creates a new [`RemoveStrategy`] instance.
    #[must_use]
    pub const fn new(strategy_id: StrategyId, command_id: UUID4, ts_init: UnixNanos) -> Self {
        Self {
            strategy_id,
            command_id,
            ts_init,
        }
    }

    #[must_use]
    pub fn as_any(&self) -> &dyn Any {
        self
    }
}

/// Commands handled by the [`Controller`](crate::controller::Controller).
#[derive(Debug, Clone)]
pub enum ControllerCommand {
    CreateActor(CreateActor),
    StartActor(StartActor),
    StopActor(StopActor),
    RemoveActor(RemoveActor),
    CreateStrategy(CreateStrategy),
    StartStrategy(StartStrategy),
    StopStrategy(StopStrategy),
    ExitMarket(StrategyId),
    RemoveStrategy(RemoveStrategy),
}

impl From<CreateActor> for ControllerCommand {
    fn from(command: CreateActor) -> Self {
        Self::CreateActor(command)
    }
}

impl From<CreateStrategy> for ControllerCommand {
    fn from(command: CreateStrategy) -> Self {
        Self::CreateStrategy(command)
    }
}

impl From<StartActor> for ControllerCommand {
    fn from(command: StartActor) -> Self {
        Self::StartActor(command)
    }
}

impl From<StartStrategy> for ControllerCommand {
    fn from(command: StartStrategy) -> Self {
        Self::StartStrategy(command)
    }
}

impl From<StopActor> for ControllerCommand {
    fn from(command: StopActor) -> Self {
        Self::StopActor(command)
    }
}

impl From<StopStrategy> for ControllerCommand {
    fn from(command: StopStrategy) -> Self {
        Self::StopStrategy(command)
    }
}

impl From<RemoveActor> for ControllerCommand {
    fn from(command: RemoveActor) -> Self {
        Self::RemoveActor(command)
    }
}

impl From<RemoveStrategy> for ControllerCommand {
    fn from(command: RemoveStrategy) -> Self {
        Self::RemoveStrategy(command)
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use rstest::rstest;

    use super::*;

    #[rstest]
    fn test_create_actor_command_fields() {
        let actor_config = ImportableActorConfig {
            actor_path: "tests.actors:Actor".to_string(),
            config_path: "tests.actors:ActorConfig".to_string(),
            config: HashMap::new(),
        };
        let command_id = UUID4::new();
        let command = CreateActor::new(actor_config.clone(), false, command_id, UnixNanos::new(1));

        assert_eq!(command.actor_config.actor_path, actor_config.actor_path);
        assert_eq!(command.actor_config.config_path, actor_config.config_path);
        assert!(!command.start);
        assert_eq!(command.command_id, command_id);
        assert_eq!(command.ts_init, UnixNanos::new(1));
        assert!(matches!(
            ControllerCommand::from(command),
            ControllerCommand::CreateActor(_)
        ));
    }

    #[rstest]
    fn test_create_actor_command_serde_round_trip() {
        let actor_config = ImportableActorConfig {
            actor_path: "tests.actors:Actor".to_string(),
            config_path: "tests.actors:ActorConfig".to_string(),
            config: HashMap::from([(
                "threshold".to_string(),
                serde_json::Value::String("10".to_string()),
            )]),
        };
        let command_id = UUID4::new();
        let command = CreateActor::new(actor_config, true, command_id, UnixNanos::new(9));

        let value = serde_json::to_value(&command).unwrap();
        assert_eq!(value["type"], "CreateActor");
        let round_trip: CreateActor = serde_json::from_value(value).unwrap();

        assert_eq!(round_trip.actor_config.actor_path, "tests.actors:Actor");
        assert_eq!(
            round_trip.actor_config.config_path,
            "tests.actors:ActorConfig"
        );
        assert_eq!(
            round_trip.actor_config.config["threshold"],
            serde_json::Value::String("10".to_string())
        );
        assert!(round_trip.start);
        assert_eq!(round_trip.command_id, command_id);
        assert_eq!(round_trip.ts_init, UnixNanos::new(9));
    }

    #[rstest]
    fn test_create_strategy_command_fields() {
        let strategy_config = ImportableStrategyConfig {
            strategy_path: "tests.strategies:Strategy".to_string(),
            config_path: "tests.strategies:StrategyConfig".to_string(),
            config: HashMap::new(),
        };
        let command_id = UUID4::new();
        let command =
            CreateStrategy::new(strategy_config.clone(), true, command_id, UnixNanos::new(2));

        assert_eq!(
            command.strategy_config.strategy_path,
            strategy_config.strategy_path
        );
        assert_eq!(
            command.strategy_config.config_path,
            strategy_config.config_path
        );
        assert!(command.start);
        assert_eq!(command.command_id, command_id);
        assert_eq!(command.ts_init, UnixNanos::new(2));
        assert!(matches!(
            ControllerCommand::from(command),
            ControllerCommand::CreateStrategy(_)
        ));
    }

    #[rstest]
    fn test_create_strategy_command_serde_round_trip() {
        let strategy_config = ImportableStrategyConfig {
            strategy_path: "tests.strategies:Strategy".to_string(),
            config_path: "tests.strategies:StrategyConfig".to_string(),
            config: HashMap::from([(
                "threshold".to_string(),
                serde_json::Value::String("20".to_string()),
            )]),
        };
        let command_id = UUID4::new();
        let command = CreateStrategy::new(strategy_config, false, command_id, UnixNanos::new(11));

        let value = serde_json::to_value(&command).unwrap();
        assert_eq!(value["type"], "CreateStrategy");
        let round_trip: CreateStrategy = serde_json::from_value(value).unwrap();

        assert_eq!(
            round_trip.strategy_config.strategy_path,
            "tests.strategies:Strategy"
        );
        assert_eq!(
            round_trip.strategy_config.config_path,
            "tests.strategies:StrategyConfig"
        );
        assert_eq!(
            round_trip.strategy_config.config["threshold"],
            serde_json::Value::String("20".to_string())
        );
        assert!(!round_trip.start);
        assert_eq!(round_trip.command_id, command_id);
        assert_eq!(round_trip.ts_init, UnixNanos::new(11));
    }

    #[rstest]
    fn test_start_actor_command_fields() {
        let command_id = UUID4::new();
        let command = StartActor::new(ActorId::from("Actor-001"), command_id, UnixNanos::new(3));

        assert_eq!(command.actor_id, ActorId::from("Actor-001"));
        assert_eq!(command.command_id, command_id);
        assert_eq!(command.ts_init, UnixNanos::new(3));
        assert!(matches!(
            ControllerCommand::from(command),
            ControllerCommand::StartActor(_)
        ));
    }

    #[rstest]
    fn test_start_actor_command_serde_round_trip() {
        let actor_id = ActorId::from("Actor-001");
        let command_id = UUID4::new();
        let command = StartActor::new(actor_id, command_id, UnixNanos::new(10));

        let value = serde_json::to_value(command).unwrap();
        assert_eq!(value["type"], "StartActor");
        let round_trip: StartActor = serde_json::from_value(value).unwrap();

        assert_eq!(round_trip.actor_id, actor_id);
        assert_eq!(round_trip.command_id, command_id);
        assert_eq!(round_trip.ts_init, UnixNanos::new(10));
    }

    #[rstest]
    fn test_stop_actor_command_fields() {
        let command_id = UUID4::new();
        let command = StopActor::new(ActorId::from("Actor-001"), command_id, UnixNanos::new(4));

        assert_eq!(command.actor_id, ActorId::from("Actor-001"));
        assert_eq!(command.command_id, command_id);
        assert_eq!(command.ts_init, UnixNanos::new(4));
        assert!(matches!(
            ControllerCommand::from(command),
            ControllerCommand::StopActor(_)
        ));
    }

    #[rstest]
    fn test_remove_actor_command_fields() {
        let command_id = UUID4::new();
        let command = RemoveActor::new(ActorId::from("Actor-001"), command_id, UnixNanos::new(5));

        assert_eq!(command.actor_id, ActorId::from("Actor-001"));
        assert_eq!(command.command_id, command_id);
        assert_eq!(command.ts_init, UnixNanos::new(5));
        assert!(matches!(
            ControllerCommand::from(command),
            ControllerCommand::RemoveActor(_)
        ));
    }

    #[rstest]
    fn test_start_strategy_command_fields() {
        let command_id = UUID4::new();
        let command = StartStrategy::new(
            StrategyId::from("Strategy-001"),
            command_id,
            UnixNanos::new(6),
        );

        assert_eq!(command.strategy_id, StrategyId::from("Strategy-001"));
        assert_eq!(command.command_id, command_id);
        assert_eq!(command.ts_init, UnixNanos::new(6));
        assert!(matches!(
            ControllerCommand::from(command),
            ControllerCommand::StartStrategy(_)
        ));
    }

    #[rstest]
    fn test_start_strategy_command_serde_round_trip() {
        let strategy_id = StrategyId::from("Strategy-001");
        let command_id = UUID4::new();
        let command = StartStrategy::new(strategy_id, command_id, UnixNanos::new(12));

        let value = serde_json::to_value(command).unwrap();
        assert_eq!(value["type"], "StartStrategy");
        let round_trip: StartStrategy = serde_json::from_value(value).unwrap();

        assert_eq!(round_trip.strategy_id, strategy_id);
        assert_eq!(round_trip.command_id, command_id);
        assert_eq!(round_trip.ts_init, UnixNanos::new(12));
    }

    #[rstest]
    fn test_stop_strategy_command_fields() {
        let command_id = UUID4::new();
        let command = StopStrategy::new(
            StrategyId::from("Strategy-001"),
            command_id,
            UnixNanos::new(7),
        );

        assert_eq!(command.strategy_id, StrategyId::from("Strategy-001"));
        assert_eq!(command.command_id, command_id);
        assert_eq!(command.ts_init, UnixNanos::new(7));
        assert!(matches!(
            ControllerCommand::from(command),
            ControllerCommand::StopStrategy(_)
        ));
    }

    #[rstest]
    fn test_remove_strategy_command_fields() {
        let command_id = UUID4::new();
        let command = RemoveStrategy::new(
            StrategyId::from("Strategy-001"),
            command_id,
            UnixNanos::new(8),
        );

        assert_eq!(command.strategy_id, StrategyId::from("Strategy-001"));
        assert_eq!(command.command_id, command_id);
        assert_eq!(command.ts_init, UnixNanos::new(8));
        assert!(matches!(
            ControllerCommand::from(command),
            ControllerCommand::RemoveStrategy(_)
        ));
    }
}
